// js-confuser/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Obfuscator", {
  enumerable: true,
  get: function get() {
    return _obfuscator["default"];
  }
});
Object.defineProperty(exports, "Template", {
  enumerable: true,
  get: function get() {
    return _template["default"];
  }
});
exports["default"] = void 0;
exports.obfuscate = obfuscate;
exports.obfuscateAST = obfuscateAST;
exports.obfuscateWithProfiler = obfuscateWithProfiler;
Object.defineProperty(exports, "presets", {
  enumerable: true,
  get: function get() {
    return _presets["default"];
  }
});
var _obfuscator = _interopRequireDefault(require("./obfuscator"));
var _presets = _interopRequireDefault(require("./presets"));
var _template = _interopRequireDefault(require("./templates/template"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function obfuscate(_x, _x2) {
  return _obfuscate.apply(this, arguments);
}
function _obfuscate() {
  _obfuscate = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(sourceCode, options) {
    var obfuscator;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.n) {
        case 0:
          obfuscator = new _obfuscator["default"](options);
          return _context.a(2, obfuscator.obfuscate(sourceCode));
      }
    }, _callee);
  }));
  return _obfuscate.apply(this, arguments);
}
function obfuscateAST(_x3, _x4) {
  return _obfuscateAST.apply(this, arguments);
}
function _obfuscateAST() {
  _obfuscateAST = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(ast, options) {
    var obfuscator;
    return _regenerator().w(function (_context2) {
      while (1) switch (_context2.n) {
        case 0:
          obfuscator = new _obfuscator["default"](options);
          return _context2.a(2, obfuscator.obfuscateAST(ast));
      }
    }, _callee2);
  }));
  return _obfuscateAST.apply(this, arguments);
}
function obfuscateWithProfiler(_x5, _x6) {
  return _obfuscateWithProfiler.apply(this, arguments);
}
function _obfuscateWithProfiler() {
  _obfuscateWithProfiler = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee3(sourceCode, options) {
    var _profiler,
      startTime,
      obfuscator,
      totalTransforms,
      transformMap,
      beforeParseTime,
      ast,
      parseTime,
      currentTransformTime,
      beforeCompileTime,
      code,
      compileTime,
      endTime,
      obfuscationTime,
      _args3 = arguments;
    return _regenerator().w(function (_context3) {
      while (1) switch (_context3.n) {
        case 0:
          _profiler = _args3.length > 2 && _args3[2] !== undefined ? _args3[2] : {};
          if (!_profiler.performance) {
            _profiler.performance = {
              now: function now() {
                return Date.now();
              }
            };
          }
          startTime = performance.now();
          obfuscator = new _obfuscator["default"](options);
          totalTransforms = obfuscator.plugins.length;
          transformMap = Object.create(null);
          beforeParseTime = performance.now();
          ast = _obfuscator["default"].parseCode(sourceCode);
          parseTime = performance.now() - beforeParseTime;
          currentTransformTime = performance.now();
          ast = obfuscator.obfuscateAST(ast, {
            profiler: function profiler(log) {
              var _profiler$callback;
              var nowTime = performance.now();
              var entry = {
                transformTime: nowTime - currentTransformTime,
                changeData: {}
              };
              transformMap[log.currentTransform] = entry;

              // (JS-Confuser.com can run performance benchmark tests here)
              (_profiler$callback = _profiler.callback) === null || _profiler$callback === void 0 || _profiler$callback.call(_profiler, log, entry, ast);

              // The profiler.callback() function may take a long time to execute,
              // so we need to update the currentTransformTime here for accurate profiling.
              currentTransformTime = performance.now();
            }
          });
          obfuscator.plugins.forEach(function (_ref) {
            var pluginInstance = _ref.pluginInstance;
            if (transformMap[pluginInstance.name]) {
              transformMap[pluginInstance.name].changeData = pluginInstance.changeData;
            }
          });
          beforeCompileTime = performance.now();
          code = _obfuscator["default"].generateCode(ast, obfuscator.options);
          compileTime = performance.now() - beforeCompileTime;
          endTime = performance.now();
          obfuscationTime = endTime - startTime;
          return _context3.a(2, {
            code: code,
            profileData: {
              transforms: transformMap,
              obfuscationTime: obfuscationTime,
              parseTime: parseTime,
              compileTime: compileTime,
              totalTransforms: totalTransforms,
              totalPossibleTransforms: obfuscator.totalPossibleTransforms
            }
          });
      }
    }, _callee3);
  }));
  return _obfuscateWithProfiler.apply(this, arguments);
}
var JsConfuser = {
  obfuscate: obfuscate,
  obfuscateAST: obfuscateAST,
  obfuscateWithProfiler: obfuscateWithProfiler,
  Obfuscator: _obfuscator["default"],
  presets: _presets["default"],
  Template: _template["default"]
};
var _default = exports["default"] = JsConfuser;
